src/content/peerconnection/negotiate-timing/js/main.js

```diff
@@ -0,0 +1,228 @@
+  pc2.onicecandidate = e => onIceCandidate(pc2, e);
+  pc1.oniceconnectionstatechange = e => onIceStateChange(pc1, e);
+  pc2.oniceconnectionstatechange = e => onIceStateChange(pc2, e);
+  pc2.addEventListener('track', gotRemoteStream, {once: true});
+
+  localStream.getTracks().forEach(track => pc1.addTrack(track, localStream));
+  console.log('Added local stream to pc1');
+
+  console.log('pc1 createOffer start');
+  const offer = await pc1.createOffer(offerOptions);
+  console.log(`Offer from pc1\n${offer.sdp}`);
+  console.log('pc1 setLocalDescription start');
+  await Promise.all([pc1.setLocalDescription(offer), pc2.setRemoteDescription(offer)]);
+  console.log(`setLocalDescription offer complete`);
+  const answer = await pc2.createAnswer();
+  await pc1.setRemoteDescription(answer);
+  await pc2.setLocalDescription(answer);
+  console.log('set*Description(answer) complete');
+}
+
+function gotRemoteStream(e) {
+  console.log('gotRemoteStream', e.track, e.streams[0]);
+  if (e.streams[0]) {
+    // reset srcObject to work around minor bugs in Chrome and Edge.
+    remoteVideo.srcObject = null;
+    remoteVideo.srcObject = e.streams[0];
+  }
+}
+
+async function onIceCandidate(pc, event) {
+  console.log(`${getName(pc)} ICE candidate:\n${event.candidate ? event.candidate.candidate : '(null)'}`);
+  await getOtherPc(pc).addIceCandidate(event.candidate);
+  console.log(`${getName(pc)} addIceCandidate success`);
+}
+
+function onIceStateChange(pc, event) {
+  if (pc) {
+    console.log(`${getName(pc)} ICE state: ${pc.iceConnectionState}`);
+    console.log('ICE state change event, state: ', pc.iceConnectionState);
+  }
+}
+
+function adjustTransceiverCounts(pc, videoCount) {
+  const currentVideoTransceivers = pc.getTransceivers().filter(tr => tr.receiver.track.kind == 'video');
+  const currentVideoCount = currentVideoTransceivers.length;
+  if (currentVideoCount < videoCount) {
+    console.log('Adding ' + (videoCount - currentVideoCount) + ' transceivers');
+    for (let i = currentVideoCount; i < videoCount; ++i) {
+      pc.addTransceiver('video');
+    }
+  } else if (currentVideoCount > videoCount) {
+    console.log('Stopping ' + (currentVideoCount - videoCount) + ' transceivers');
+    for (let i = videoCount; i < currentVideoCount; ++i) {
+      currentVideoTransceivers[i].stop();
+    }
+  } else {
+    console.log(`No adjustment, video count is ${currentVideoCount}, target was ${videoCount}`);
+  }
+}
+
+async function getAudioImpairment(audioTransceiver) {
+  const stats = await audioTransceiver.receiver.getStats();
+  let currentImpairment;
+  stats.forEach(stat => {
+    if (stat.type == 'track') {
+      currentImpairment = stat.concealedSamples;
+    }
+  });
+  console.log('Found impairment value ', currentImpairment);
+  return currentImpairment;
+}
+
+async function baselineAudioImpairment(pc) {
+  audioTransceiver = pc.getTransceivers().filter(tr => tr.receiver.track.kind == 'audio')[0];
```

Comment:
.find and remove the [0]. Should aduioTransceiver be set in ontrack?

Reply:
Could do that, but then ontrack has to distinguish between 3 cases (audio, video and "don't care").

Reply:
Maybe doc that baselineAudioImpairment must be called before measureAudioImpairment and/or that they should be called in pairs.